util/oci/client.go

```diff
@@ -0,0 +1,544 @@
+func extractContentToManifestsDir(ctx context.Context, cachedPath, digest string, maxSize int64, allowedMediaTypes []string) (string, error) {
+	manifestsDir, err := files.CreateTempDir(os.TempDir())
+	if err != nil {
+		return manifestsDir, err
+	}
+
+	ociReadOnlyStore, err := oci.NewFromTar(ctx, cachedPath)
+	if err != nil {
+		return manifestsDir, err
+	}
+
+	tempDir, err := files.CreateTempDir(os.TempDir())
+	if err != nil {
+		return manifestsDir, err
+	}
+	defer os.RemoveAll(tempDir)
+
+	fs, err := newCompressedLayerFileStore(manifestsDir, tempDir, maxSize, allowedMediaTypes)
+	if err != nil {
+		return manifestsDir, err
+	}
+	defer fs.Close()
+
+	// copies the whole artifact to the tempdir, here compressedLayerFileStore.Push will be called
+	_, err = oras.Copy(ctx, ociReadOnlyStore, digest, fs, digest, oras.DefaultCopyOptions)
+	return manifestsDir, err
+}
+
+type compressedLayerExtracterStore struct {
+	*file.Store
+	tempDir           string
+	dest              string
+	maxSize           int64
+	allowedMediaTypes []string
+}
+
+func newCompressedLayerFileStore(dest, tempDir string, maxSize int64, allowedMediaTypes []string) (*compressedLayerExtracterStore, error) {
+	f, err := file.New(tempDir)
+	if err != nil {
+		return nil, err
+	}
+
+	return &compressedLayerExtracterStore{f, tempDir, dest, maxSize, allowedMediaTypes}, nil
+}
+
+// Push looks in all the layers of an OCI image. Once it finds a layer that is compressed, it extracts the layer to a tempDir
+// and then renames the temp dir to the directory where the repo-server expects to find k8s manifests.
+func (s *compressedLayerExtracterStore) Push(ctx context.Context, desc v1.Descriptor, content io.Reader) error {
+	if isCompressedLayer(desc.MediaType) {
+		tempDir, err := files.CreateTempDir(os.TempDir())
+		if err != nil {
+			return err
+		}
+		defer os.RemoveAll(tempDir)
+
+		if strings.HasSuffix(desc.MediaType, "tar+gzip") {
+			err = files.Untgz(tempDir, content, s.maxSize, false)
+		} else {
+			err = files.Untar(tempDir, content, s.maxSize, false)
+		}
+
+		if err != nil {
+			return fmt.Errorf("could not decompress layer: %w", err)
+		}
+
+		infos, err := os.ReadDir(tempDir)
+		if err != nil {
+			return err
+		}
+
+		if isHelmOCI(desc.MediaType) {
+			// For a Helm chart we expect a single tarfile in the directory
+			if len(infos) != 1 {
+				return fmt.Errorf("expected 1 file, found %v", len(infos))
+			}
+		}
+
+		if len(infos) == 1 && infos[0].IsDir() {
+			// Here we assume that this is a directory which has been decompressed. We need to move the contents of
+			// the dir into our intended destination.
+			srcDir := filepath.Join(tempDir, infos[0].Name())
+			return filepath.WalkDir(srcDir, func(path string, d fs.DirEntry, err error) error {
+				if path != srcDir {
+					// Calculate the relative path from srcDir
+					relPath, err := filepath.Rel(srcDir, path)
+					if err != nil {
+						return err
+					}
+
+					dstPath := filepath.Join(s.dest, relPath)
```

Comment:
Ditto re: securejoin. If `relPath` refers to a symlink, this join could end up referring to something above `s.dest`.